### PR TITLE
Feature : Enable Intel GPU/XPU finetuning and inference

### PR DESCRIPTION
**Pull request by abhilash1910**
# What does this PR do?

Motivation:  Thanks for creating this repository. At Intel, we are using this repository for finetuning and inference for our GPUs /XPU (Ipex backend) and hence the reason for this PR. 
This change would help us run LLama v2 without issues on our GPU devices. (Since accelerate/peft already has support of XPUs enabled)

Fixes # (issue) -NA

## Feature/Issue validation/testing

Feature for finetuning and inference tested on our Data Centre GPUs with the change

for CLA purposes this would be on behalf of Intel. 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/llama-recipes/blob/main/CONTRIBUTING.md#pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes?  
- [ ] Did you write any new necessary tests?

Thanks for contributing 🎉!

### CODE REVIEW

**Comment by HamidShojanazeri** [comment]
@abhilash1910 thanks the PR, very interesting, would love to learn more about how we can test Intel GPUs/XPUs. I sent you a message on LD.

**Comment by chauhang** [comment]
@abhilash1910 Thanks for submitting this PR. Please attach the logs for the tests for verification that things are working on the Intel GPUs with the changes. We don't have a mechanism to test this. 

Can you also please check if [torch.set_default_device](https://pytorch.org/docs/stable/generated/torch.set_default_device.html) api can be used to simply the device setting logic for the XPU devices?

**Comment by abhilash1910** [comment]
@chauhang  sure I will attach the logs; also as discussed with @HamidShojanazeri  there is a plan to support device for integration testing for any future PRs from Intel side. I am closing on it this week internally , and eventually would like to discuss on the next steps.

File: utils/train_utils.py [hunk 1]
```diff
@@ -125,17 +131,29 @@ def train(model, train_dataloader,eval_dataloader, tokenizer, optimizer, lr_sche
         
         if train_config.enable_fsdp:
             if rank==0:
+                if is_xpu_available():
```
Comment (maximegmd) [inline utils/train_utils.py:134]: This code could be simplified`:
```python
device_kind = "CPU" if is_xpu_available() else "CUDA"
print(f"Max {device_kind} memory allocated was {memtrace.peak} GB")
print(f"Max {device_kind} memory reserved was {memtrace.max_reserved} GB")
print(f"Peak active {device_kind} memory was {memtrace.peak_active_gb} GB")
print(f"{device_kind} Malloc retires : {memtrace.cuda_malloc_retires}")

File: utils/train_utils.py [hunk 2]
```diff
@@ -125,17 +131,29 @@ def train(model, train_dataloader,eval_dataloader, tokenizer, optimizer, lr_sche
...
+                print(f"Max XPU memory allocated was {memtrace.peak} GB")
+                print(f"Max XPU memory reserved was {memtrace.max_reserved} GB")
+                print(f"Peak active XPU memory was {memtrace.peak_active_gb} GB")
+                print(f"Xpu Malloc retires : {memtrace.cuda_malloc_retires}")
+            else:
                 print(f"Max CUDA memory allocated was {memtrace.peak} GB")
                 print(f"Max CUDA memory reserved was {memtrace.max_reserved} GB")
                 print(f"Peak active CUDA memory was {memtrace.peak_active_gb} GB")
                 print(f"Cuda Malloc retires : {memtrace.cuda_malloc_retires}")
                 print(f"CPU Total Peak Memory consumed during the train (max): {memtrace.cpu_peaked + memtrace.cpu_begin} GB")
```
Comment (maximegmd) [inline utils/train_utils.py:156]: I think this line should be unindented

**Comment by abhilash1910** [comment]
@chauhang @HamidShojanazeri logs for distributed finetuning on 8 ranks :
[7B_8ranks_intel_xpu.log](https://github.com/facebookresearch/llama-recipes/files/12519127/7B_8ranks_intel_xpu.log)
Could you also share a tentative timing for a discussion sometime this week or next on possible next steps?

**Comment by HamidShojanazeri** [comment]
Thanks very much @abhilash1910, I am wondering if there is a way to add support/ get intel GPU memory stats as well the way we are getting for nvidia.

**Comment by abhilash1910** [comment]
@HamidShojanazeri yes the memory stats are also added , this can be found in the new log here:
[7B_8ranks_intel_xpu.log](https://github.com/facebookresearch/llama-recipes/files/12544371/7B_8ranks_intel_xpu.log)
Thanks

**Review by HamidShojanazeri — APPROVED** [review]

File: src/llama_recipes/utils/train_utils.py [hunk 3]
```diff
@@ -21,7 +21,9 @@
 from llama_recipes.model_checkpointing import save_model_checkpoint, save_model_and_optimizer_sharded, save_optimizer_checkpoint
 from llama_recipes.policies import fpSixteen,bfSixteen_mixed, get_llama_wrapper
 from llama_recipes.utils.memory_utils import MemoryTrace
+from accelerate.utils import is_xpu_available, is_ccl_available
 
+from accelerate.utils import is_xpu_available, is_ccl_available
```
Comment (gujinghui) [inline src/llama_recipes/utils/train_utils.py:26]: duplicated import?

Comment (abhilash1910) [reply src/llama_recipes/utils/train_utils.py]: thanks @gujinghui

**Pull request merged**